internal/server/boltdbstate/project.go

```diff
@@ -69,10 +69,185 @@ func (s *State) ProjectGet(ref *pb.Ref_Project) (*pb.Project, error) {
+				// Get the triggers for a project in the workspace
+				if triggerList, err := s.TriggerList(workspace.Workspace, &pb.Ref_Project{Project: project.Name}, nil, []string{}); err != nil {
+					return err
+				} else {
+					triggers = append(triggers, triggerList...)
+				}
+				if workspaceDetail, err := s.WorkspaceGet(workspace.Workspace.Workspace); err != nil {
+					return err
+				} else {
+					// If the project we're deleting is the only project in the workspace, we delete the workspace
+					// We don't delete the default workspace
+					if len(workspaceDetail.Projects) == 1 && workspace.Workspace.Workspace != "default" {
+						workspaces = append(workspaces, workspace)
+					}
+				}
+
+			}
+		}
+
+		if pipelines, err = s.PipelineList(ref); err != nil {
+			return err
+		}
+
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	// Builds, artifacts, deployments, releases, status reports, pipelines, triggers,
+	// workspaces and config are deleted with a project
+	// Jobs and tasks will NOT be deleted along with a project
+	// Instances are expected to be deleted before ProjectDelete, via the destroy op
+	for _, build := range builds {
+		if err = s.BuildDelete(&pb.Ref_Operation{Target: &pb.Ref_Operation_Id{Id: build.Id}}); err != nil {
+			return err
+		}
+		s.log.Debug("deleted build " + build.Id)
+	}
+	for _, artifact := range artifacts {
+		if err = s.ArtifactDelete(&pb.Ref_Operation{Target: &pb.Ref_Operation_Id{Id: artifact.Id}}); err != nil {
+			return err
+		}
+	}
+	for _, deployment := range deployments {
+		if err = s.DeploymentDelete(&pb.Ref_Operation{Target: &pb.Ref_Operation_Id{Id: deployment.Id}}); err != nil {
+			return err
+		}
+	}
+	for _, release := range releases {
+		if err = s.ReleaseDelete(&pb.Ref_Operation{Target: &pb.Ref_Operation_Id{Id: release.Id}}); err != nil {
+			return err
+		}
+	}
+	for _, statusReport := range statusReports {
+		if err = s.StatusReportDelete(&pb.Ref_Operation{Target: &pb.Ref_Operation_Id{Id: statusReport.Id}}); err != nil {
+			return err
+		}
+	}
+
+	// Unset all configs we retrieved
+	for _, config := range configVars {
+		if err = s.ConfigSet(&pb.ConfigVar{
+			Target:     config.Target,
+			Name:       config.Name,
+			Value:      &pb.ConfigVar_Unset{},
+			Internal:   config.Internal,
+			NameIsPath: config.NameIsPath,
+		}); err != nil {
+			return err
+		}
+	}
+
+	// delete workspaces for a project
+	for _, workspace := range workspaces {
+		if err = s.WorkspaceDelete(workspace.Workspace.Workspace); err != nil {
+			return err
+		}
+	}
+
+	// delete triggers for project
+	for _, trigger := range triggers {
+		if err = s.TriggerDelete(&pb.Ref_Trigger{Id: trigger.Id}); err != nil {
+			return err
+		}
+	}
+
+	// delete pipelines for project
+	for _, pipeline := range pipelines {
+		if err = s.PipelineDelete(&pb.Ref_Pipeline{Ref: &pb.Ref_Pipeline_Id{
+			Id: &pb.Ref_PipelineId{Id: pipeline.Id},
+		},
+		}); err != nil {
+			return err
+		}
+	}
+
 	memTxn := s.inmem.Txn(true)
```

Comment:
I'm concerned that if half way through the cascading delete something fails, e.g. we delete all the configs, but fail to delete the pipelines, there is no rollback and the project is in a half-deleted state. Since it doesn't look like the cascading deletes are wrapped in a transaction. I also didn't look too closely at the ordering, but looks like workspace is deleted before project is deleted, but shouldn't that happen after?

Reply:
@xiaolin-ninja I think your comment addressed each concern that I had as I was attempting to implement this 😄 I agree it's not ideal that it's seemingly unable to be wrapped in a single transaction which can be rolled back. Fortunately though, the project is the last thing that is deleted, so if a failure happens before the project gets deleted, another attempt can be made to delete the project, and the remaining records will still be able to be referenced by the project.

Regarding the comment about `workspace`, I think it's okay to delete the workspace before the project, because if the project is destroyed but we fail to delete the workspace after that, then we won't be able to later look up the workspace(s) by project, since the project will be gone.

Reply:
All of these should be called in a single `db.Update()` with the lower-case state versions used (meant to called inside a transaction, i.e. `triggerDelete` instead of `TriggerDelete`.). Then if _any_ of them fail, the entire transaction is rolled back. Doing all of this in a single transaction means we shouldn't have any issues with race conditions or locking as far as I know.

Reply:
Updated! 😎